server/src/main/java/org/elasticsearch/index/codec/vectors/ES814ScalarQuantizedVectorsWriter.java

```diff
@@ -0,0 +1,1061 @@
+import org.elasticsearch.vec.VectorScorerProvider;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.lucene.codecs.lucene99.Lucene99ScalarQuantizedVectorsFormat.QUANTIZED_VECTOR_COMPONENT;
+import static org.apache.lucene.codecs.lucene99.Lucene99ScalarQuantizedVectorsFormat.calculateDefaultConfidenceInterval;
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.apache.lucene.util.RamUsageEstimator.shallowSizeOfInstance;
+
+/**
+ * Writes quantized vector values and metadata to index segments.
+ * Amended copy of Lucene99ScalarQuantizedVectorsWriter
+ */
+public final class ES814ScalarQuantizedVectorsWriter extends FlatVectorsWriter {
+
+    static final int DIRECT_MONOTONIC_BLOCK_SHIFT = 16;
+
+    private static final long SHALLOW_RAM_BYTES_USED = shallowSizeOfInstance(ES814ScalarQuantizedVectorsWriter.class);
+
+    // Used for determining when merged quantiles shifted too far from individual segment quantiles.
+    // When merging quantiles from various segments, we need to ensure that the new quantiles
+    // are not exceptionally different from an individual segments quantiles.
+    // This would imply that the quantization buckets would shift too much
+    // for floating point values and justify recalculating the quantiles. This helps preserve
+    // accuracy of the calculated quantiles, even in adversarial cases such as vector clustering.
+    // This number was determined via empirical testing
+    private static final float QUANTILE_RECOMPUTE_LIMIT = 32;
+    // Used for determining if a new quantization state requires a re-quantization
+    // for a given segment.
+    // This ensures that in expectation 4/5 of the vector would be unchanged by requantization.
+    // Furthermore, only those values where the value is within 1/5 of the centre of a quantization
+    // bin will be changed. In these cases the error introduced by snapping one way or another
+    // is small compared to the error introduced by quantization in the first place. Furthermore,
+    // empirical testing showed that the relative error by not requantizing is small (compared to
+    // the quantization error) and the condition is sensitive enough to detect all adversarial cases,
+    // such as merging clustered data.
+    private static final float REQUANTIZATION_LIMIT = 0.2f;
+    private final SegmentWriteState segmentWriteState;
+
+    private final List<FieldWriter> fields = new ArrayList<>();
+    private final IndexOutput meta, quantizedVectorData;
+    private final Float confidenceInterval;
+    private final FlatVectorsWriter rawVectorDelegate;
+    private boolean finished;
+
+    ES814ScalarQuantizedVectorsWriter(SegmentWriteState state, Float confidenceInterval, FlatVectorsWriter rawVectorDelegate)
```

Comment:
I think we could reduce the amount of code copying if `ES814ScalarQuantizedVectorsWriter` constructed a `Lucene99ScalarQuantizedVectorsWriter` and used it internally for all the methods. We should be able to defer a bunch of logic to it.

Reply:
Yeah, I had the same thought too, but it won't work :-(  The reason why it will not work is that `Lucene99ScalarQuantizedVectorsWriter` has too much internal state that it keeps private. An external "wrapper" is not able to gain access to, say, the temporary file where the vector data lives.